adds tracking flags to the tracked pose driver.

mattdatwork opened the pull request:
### Description

Adds optional tracking state flags to the Tracked Pose Driver.
When no action reference / inline action to the tracking state flags (int) is used, the Tracked Pose Driver operates as per current.
When an action is provided, we check the flags and only apply the position/rotation components that were actually indicated as correct by the tracking state flags.

This will resolve snap to zero issues when using the TPD

### Changes made

Change the trackedposedriver.cs file to allow additional Tracking State Flag actions.


### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

MattDalby committed:
adds tracking flags to the tracked pose driver.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1481) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1481) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/InputSystem?pullRequest=1481) it.</sub>

chris-massie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -320,19 +354,40 @@ protected virtual void OnBeforeRender()
         }
 
-        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation)
+        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation, bool hasTrackingStateFlagInfo, PoseDataFlags trackingStateFlags)
```
This is a `protected virtual` method, so you can't just update the signature without it being a breaking change for existing users since their code wouldn't compile.

mattdatwork replied:
good call. adding a facade call through.

chris-massie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -135,11 +135,45 @@ public InputActionProperty rotationInput
+        /// </summary>
+
+        [Flags]
+        protected enum PoseDataFlags
```
Should `UnityEngine.XR.InputTrackingState` from the XR Module be used instead since that's the actual tracking enum being used? That would require adding a dependency on the XR Module, but then users would have access to all of the information.

chris-massie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -135,11 +135,45 @@ public InputActionProperty rotationInput
+
         Vector3 m_CurrentPosition = Vector3.zero;
         Quaternion m_CurrentRotation = Quaternion.identity;
+        bool m_HasTrackingStateFlags = false;
```
There's no reason to have this as a class field since it is only used locally.

vrdave-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -320,19 +354,40 @@ protected virtual void OnBeforeRender()
+            if (hasTrackingStateFlagInfo)
             {
-                transform.localRotation = newRotation;
+                if ((trackingStateFlags & PoseDataFlags.Position) == PoseDataFlags.Position &&
```
Is this supposed to be `PoseDataFlags.Rotation` (swap logic with line 368)?

vrdave-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -320,19 +354,40 @@ protected virtual void OnBeforeRender()
+                    transform.localRotation = newRotation;
+                }
+
+                if ((trackingStateFlags & PoseDataFlags.Rotation) == PoseDataFlags.Rotation &&
```
(See comment for line 361)

vrdave-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -320,19 +354,40 @@ protected virtual void OnBeforeRender()
         }
 
-        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation)
+        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation, bool hasTrackingStateFlagInfo, PoseDataFlags trackingStateFlags)
```
This method signature doesn't look like it has to change since you are using member variables to track `m_HasTrackingStateFlags` and `m_CurrentTrackingStateFlags`, correct? Are there other consumers of this function that will change the transform data without changing the 'new' internal state of this object?

caryhaynie replied:
Two things:
* Can't `hasTrackingStateInfo` be derived from `trackingStateFlags`? (eg, `hasTrackingStateInfo = trackingStateFlags != PoseDataFlags.None;`
* You could also just overload the method and have to 2-param version call into the 4-param version (using the member variables @vrdave-unity alluded to above to supply the additional parameters?)

mattdatwork replied:
added a pass through.

and i need the additional flag because i want tracking to work if there's no action providing tracking state flags. a None will cause nothing to be applied.

bryandube reviewed (approved):
Aside from being sad that we adding another possible inline action it looks like the other reviewers caught everything that I noticed already.

caryhaynie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -343,7 +398,17 @@ bool HasStereoCamera()
 
         protected virtual void PerformUpdate()
```
(I realize this isn't part of this actual PR, but I wanted to ask about it)
Why is this method virtual? Is there case where the user will want to completely ignore this behavior and do something entirely different instead? Additionally, because this method uses private variables, isn't it functionally impossible for a user to implement a replacement correctly?

mattdatwork replied:
just allows us to call the same code in update/onbefore render. figured it might be nice for people to override this. but yeah. lack of overall protected et al is probably not great.

caryhaynie commented:
> Aside from being sad that we adding another possible inline action ...

👍 for a new yamato job: "Make sure @bryandube is not saddened by this change." 🤣

MattDalby committed:
pr requested changes.

caryhaynie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -354,18 +331,24 @@ protected virtual void OnBeforeRender()
 
-        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation, bool hasTrackingStateFlagInfo, PoseDataFlags trackingStateFlags)
+
+        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation)
```
probably want to add an [Obsolete] attribute here, pointing people to the new override.

caryhaynie commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -1,5 +1,6 @@
 using System;
 using UnityEngine.InputSystem.LowLevel;
+using UnityEngine.XR;
```
as @chris-massie mentioned, need to add a dependency on `com.unity.modules.xr` for this.

caryhaynie replied:
(in package.json, just to be clear)

Rene-Damm reviewed (changes requested):
Missing tests and changelog.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -343,7 +381,19 @@ bool HasStereoCamera()
         {
-            SetLocalTransform(m_CurrentPosition, m_CurrentRotation);
+            InputTrackingState currentTrackingStateFlags = InputTrackingState.None;
+            bool hasTrackingStateFlags = false;
```
Minor, but why not direct evaluate here to reduce this code? E.g.
`bool hasTrackingStateFlags = m_TrackingStateFlags?.action?.bindings.Count > 0;
var currentTrackingStateFlags = hasTrackingStateFlags ? (InputTrackingState)m_TrackingStateFlags.action.ReadValue<int>() : InputTrackingState.None;`
or even evaluate in call to SetLocalTransform and skip the locals?

ekcoh replied:
Maybe easier to debug this way....

ekcoh reviewed (commented):
Looks good to me, missing changelog/tests though similar to what was already highlighted by @Rene-Damm

MattDalby committed:
adds unit tests
re-writes handling of tracking state to follow action based approach as it would not play nice with unit tests.

MattDalby committed:
adds changelog

vrdave-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -342,8 +446,17 @@ bool HasStereoCamera()
-        {
-            SetLocalTransform(m_CurrentPosition, m_CurrentRotation);
+        {            
+            PoseDataFlags trackingStateFlags = m_CurrentTrackingStateFlags;
```
I suggest changing the name of this local variable to `updatedTrackingStateFlags` to be a bit more specific and not conflict with the globally scoped variable name.

vrdave-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -342,8 +446,17 @@ bool HasStereoCamera()
-            SetLocalTransform(m_CurrentPosition, m_CurrentRotation);
+        {            
+            PoseDataFlags trackingStateFlags = m_CurrentTrackingStateFlags;
+            if (m_TrackingStateFlags != null && m_TrackingStateFlags.action != null && m_TrackingStateFlags.action.bindings.Count > 0)
+            { 
+            }   
+            else
+            {
+                trackingStateFlags = PoseDataFlags.Position | PoseDataFlags.Rotation;
+            }
```
I feel like the following is a bit clearer and simpler for readability: 
```
if (m_TrackingStateFlags == null || m_TrackingStateFlags.action == null || m_TrackingStateFlags.action.bindings.Count == 0)
    trackingStateFlags = PoseDataFlags.Position | PoseDataFlags.Rotation;
```

Is there a performance advantage to the existing code that is relevant?

unity-andrewc commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ however, it has to be formatted properly to pass verification tests.
 - Deprecated `InputSettings.filterNoiseOnCurrent`. Now noise filtering is always enabled. Device only will become `.current` if any non-noise control have changed state.
 - A device reset (such as when focus is lost) on `Touchscreen` will now result in all ongoing touches getting cancelled instead of all touches being simply reset to default state.
 - Calling `InputTestFixture.Press`, `InputTestFixture.Set`, etc. from within a `[UnityTest]` will no longer immediately process input. Instead, input will be processed like it normally would as part of the Unity player loop.
+- Added `TrackingStateFlags` to the tracked pose driver packaged with the input system. Assigning this to an IntegerControl which provides TrackingState information will now correctly only apply those elements that actually recieved reliable data. If a TrackingStateFlags is not provided, it is assumed both position and rotation are valid.
```
recieved -> received

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -80,6 +81,28 @@ public enum UpdateType
+            /// If this flag is set, rotation data was updated on the associated pose struct
+            /// </summary>
+            Rotation = 1 << 1,
+        };
```
Remove semicolon

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -135,10 +158,30 @@ public InputActionProperty rotationInput
+        public InputActionProperty trackingStateFlags
+        {
+            get => m_TrackingStateFlags;
+            set {
```
Move opening curly brace to new line

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -135,10 +158,30 @@ public InputActionProperty rotationInput
+
+                m_TrackingStateFlags = value;
+
+                if (Application.isPlaying && isActiveAndEnabled)
```
Are there any side effects with assuming that we should bind tracking state if this line is true? Should we instead and with a local bool that we set if we unbound the tracking state above?

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -135,10 +158,30 @@ public InputActionProperty rotationInput
+
         Vector3 m_CurrentPosition = Vector3.zero;
         Quaternion m_CurrentRotation = Quaternion.identity;
+        PoseDataFlags m_CurrentTrackingStateFlags = 0;
```
`0` -> `PoseDataFlags.None`

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -266,7 +360,7 @@ protected virtual void Awake()
         /// This function is called when the object becomes enabled and active.
         /// </summary>
-        protected void OnEnable()
+        protected void OnEnable() 
```
Undo whitespace change

unity-andrewc commented on Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs:
```diff
@@ -320,20 +414,30 @@ protected virtual void OnBeforeRender()
+            SetLocalTransform(newPosition, newRotation, PoseDataFlags.Position | PoseDataFlags.Rotation);
+        }
+
+        protected virtual void SetLocalTransform(Vector3 newPosition, Quaternion newRotation,PoseDataFlags trackingStateFlags)
```
Missing space between final comma and `PoseDataFlags`

unity-andrewc reviewed (changes requested):
There's a lot of unnecessary new-lines in this PR (two consecutive line breaks, empty lines between closing curly braces) that would be nice to remove as well

chris-massie commented:
This PR should be closed and replaced with the implementation done in https://github.com/Unity-Technologies/InputSystem/pull/1594

ekcoh commented:
PR referenced by @chris-massie https://github.com/Unity-Technologies/InputSystem/pull/1594 have been merged so closing this PR since that is obsolete now.
